conformance/client/src/main/kotlin/com/connectrpc/conformance/client/Client.kt

```diff
@@ -49,56 +47,104 @@ import java.security.spec.PKCS8EncodedKeySpec
+        private const val IDEMPOTENT_UNARY_METHOD_NAME = "IdempotentUnary"
+        private const val UNIMPLEMENTED_METHOD_NAME = "Unimplemented"
+        private const val CLIENT_STREAM_METHOD_NAME = "ClientStream"
+        private const val SERVER_STREAM_METHOD_NAME = "ServerStream"
+        private const val BIDI_STREAM_METHOD_NAME = "BidiStream"
+
+        private const val UNARY_REQUEST_NAME = "connectrpc.conformance.v1.UnaryRequest"
+        private const val IDEMPOTENT_UNARY_REQUEST_NAME = "connectrpc.conformance.v1.IdempotentUnaryRequest"
+        private const val UNIMPLEMENTED_REQUEST_NAME = "connectrpc.conformance.v1.UnimplementedRequest"
+    }
+
     suspend fun handle(req: ClientCompatRequest): ClientResponseResult {
-        val invoker = invokerFactory(getProtocolClient(req))
-        val service = req.service.orEmpty()
-        if (service != "connectrpc.conformance.v1.ConformanceService") {
-            throw RuntimeException("service $service is not known")
-        }
+        val (httpClient, protocolClient) = getClient(req)
+        try {
+            val invoker = invokerFactory(protocolClient)
+            val service = req.service
+            if (service != CONFORMANCE_SERVICE_NAME) {
+                throw RuntimeException("service $service is not known")
+            }
 
-        return when (val method = req.method.orEmpty()) {
-            "Unary" -> handleUnary(invoker.unaryClient(), req)
-            // TODO: IdempotentUnary
-            "Unimplemented" -> handleUnary(invoker.unimplementedClient(), req)
-            "ClientStream" -> handleClient(invoker.clientStreamClient(), req)
-            "ServerStream" -> handleServer(invoker.serverStreamClient(), req)
-            "BidiStream" -> if (req.streamType == StreamType.STREAM_TYPE_FULL_DUPLEX_BIDI_STREAM) {
-                handleFullDuplexBidi(invoker.bidiStreamClient(), req)
-            } else {
-                handleHalfDuplexBidi(invoker.bidiStreamClient(), req)
+            return when (req.method) {
+                UNARY_METHOD_NAME -> handleUnary(invoker.unaryClient(), UNARY_REQUEST_NAME, req)
+                IDEMPOTENT_UNARY_METHOD_NAME -> handleUnary(invoker.idempotentUnaryClient(), IDEMPOTENT_UNARY_REQUEST_NAME, req)
+                UNIMPLEMENTED_METHOD_NAME -> handleUnary(invoker.unimplementedClient(), UNIMPLEMENTED_REQUEST_NAME, req)
+                CLIENT_STREAM_METHOD_NAME -> handleClient(invoker.clientStreamClient(), req)
+                SERVER_STREAM_METHOD_NAME -> handleServer(invoker.serverStreamClient(), req)
+                BIDI_STREAM_METHOD_NAME -> handleBidi(invoker.bidiStreamClient(), req)
+                else -> throw RuntimeException("method ${req.method} is not known")
             }
-            else -> throw RuntimeException("method $method is not known")
+        } finally {
+            // Clean-up HTTP client.
+            httpClient.connectionPool.evictAll()
+            httpClient.dispatcher.executorService.shutdown()
```

Comment:
I made these changes so that the conformance client can cleanly shutdown without using `System.exit`. It turns out that the old code was creating lots of HTTP clients and never cleaning them up, which left non-daemon threads around, preventing the JVM from promptly terminating after the `main` function returned.